File: py/src/schemapb/value.py
"""The single conversion point between the value worlds.

wire   — generated Value / StructValue (betterproto2 dataclasses)
native — natural Python values the engine and CEL operate on:

    Null -> None            Bool -> bool
    Int32/64, UInt32/64 -> int (honest, arbitrary precision)
    Float/Double -> float   String -> str      Bytes -> bytes
    Duration -> datetime.timedelta   Timestamp -> datetime.datetime (UTC)
    List -> list            Object/Map -> dict

The declared field kind — not the runtime type — picks the wire variant at
the boundary (canonical_value).
"""

from __future__ import annotations

import datetime as dt
import math
from typing import TypeAlias, Union, cast

import betterproto2

from schemapb._gen.schemapb import (
    ListValue,
    NullValue,
    Schema,
    SchemaField,
    StructValue,
    Value,
)

# Recursive alias: the container members stay as forward references (they
# must — the alias refers to itself), hence the quoted members.
Native: TypeAlias = Union[  # noqa: RUF036 - None leads: mirrors the wire variant order
    None,
    bool,
    int,
    float,
    str,
    bytes,
    dt.timedelta,
    dt.datetime,
    "list[Native]",
    "dict[str, Native]",
]
NativeStruct: TypeAlias = "dict[str, Native]"

_INT32_MIN = -(2**31)
_INT32_MAX = 2**31 - 1
_UINT32_MAX = 2**32 - 1


# =============================================================================
# Constructors (wire values)
# =============================================================================


def null_v() -> Value:
    return Value(null_value=NullValue.NULL_VALUE)


def bool_v(v: bool) -> Value:
    return Value(bool_value=v)


def int32_v(v: int) -> Value:
    return Value(int32_value=v)


def int64_v(v: int) -> Value:
    return Value(int64_value=v)


def uint32_v(v: int) -> Value:
    return Value(uint32_value=v)


def uint64_v(v: int) -> Value:
    return Value(uint64_value=v)


def float_v(v: float) -> Value:
    return Value(float_value=v)


def double_v(v: float) -> Value:
    return Value(double_value=v)


def str_v(v: str) -> Value:
    return Value(string_value=v)


def bytes_v(v: bytes) -> Value:
    return Value(bytes_value=v)


def duration_v(v: dt.timedelta) -> Value:
    return Value(duration_value=v)


def timestamp_v(v: dt.datetime) -> Value:
    return Value(timestamp_value=v)


def list_v(*items: Value) -> Value:
    return Value(list_value=ListValue(items=list(items)))


def struct_v(fields: dict[str, Value]) -> Value:
    return Value(struct_value=StructValue(fields=fields))


# =============================================================================
# Wire -> native
# =============================================================================


def to_native(v: Value | None) -> Native:
    if v is None:
        return None
    name, val = betterproto2.which_one_of(v, "kind")
    match name:
        case "" | "null_value":
            return None
        case "list_value":
            lv = v.list_value
            return [to_native(it) for it in (lv.items if lv is not None else [])]
        case "struct_value":
            return struct_to_native(v.struct_value)
        case "float_value":
            # float32 wire value: keep the float32-rounded double.
            return float(cast("float", val))
        case _:
            return cast("Native", val)


def struct_to_native(s: StructValue | None) -> NativeStruct:
    if s is None:
        return {}
    return {name: to_native(v) for name, v in s.fields.items()}


# =============================================================================
# Native -> wire (best fit, no schema)
# =============================================================================


def from_native(x: Native) -> Value:
    """Best-fitting variant: int -> int64, float -> double."""
    match x:
        case None:
            return null_v()
        case bool():
            return bool_v(x)
        case int():
            return int64_v(x)
        case float():
            return double_v(x)
        case str():
            return str_v(x)
        case bytes():
            return bytes_v(x)
        case dt.timedelta():
            return duration_v(x)
        case dt.datetime():
            return timestamp_v(x)
        case list():
            return list_v(*(from_native(el) for el in x))
        case dict():
            return struct_v({name: from_native(el) for name, el in x.items()})
    msg = f"schemapb: cannot convert {type(x).__name__} to Value"
    raise TypeError(msg)


def struct_from_native(m: NativeStruct) -> StructValue:
    return StructValue(fields={name: from_native(v) for name, v in m.items()})


# =============================================================================
# Numeric coercion helpers
# =============================================================================


def as_int(x: Native) -> int | None:
    """Signed integer view; floats only when integral, bools never."""
    if isinstance(x, bool):
        return None
    if isinstance(x, int):
        return x
    if isinstance(x, float) and math.isfinite(x) and x == math.trunc(x):
        return int(x)
    return None


def as_uint(x: Native) -> int | None:
    n = as_int(x)
    return n if n is not None and n >= 0 else None


def as_float(x: Native) -> float | None:
    if isinstance(x, bool):
        return None
    if isinstance(x, (int, float)):
        return float(x)
    return None


def is_native_struct(x: Native) -> bool:
    return isinstance(x, dict)


# =============================================================================
# Canonical form (field kind -> exact wire variant)
# =============================================================================


class CanonicalError(ValueError):
    """A value that cannot represent the declared kind."""


def _fail(msg: str) -> Value:
    raise CanonicalError(msg)


def canonical_value(f: SchemaField, x: Native) -> Value:
    if x is None:
        return null_v()
    if f.float is not None:
        n = as_float(x)
        return _fail(f"field {f.name}: not numeric") if n is None else float_v(n)
    if f.double is not None:
        n = as_float(x)
        return _fail(f"field {f.name}: not numeric") if n is None else double_v(n)
    if f.int32 is not None:
        n = as_int(x)
        if n is None or not _INT32_MIN <= n <= _INT32_MAX:
            return _fail(f"field {f.name}: does not fit int32")
        return int32_v(n)
    if f.int64 is not None:
        n = as_int(x)
        return _fail(f"field {f.name}: not an integer") if n is None else int64_v(n)
    if f.uint32 is not None:
        n = as_uint(x)
        if n is None or n > _UINT32_MAX:
            return _fail(f"field {f.name}: does not fit uint32")
        return uint32_v(n)
    if f.uint64 is not None:
        n = as_uint(x)
        return _fail(f"field {f.name}: not an unsigned integer") if n is None else uint64_v(n)
    if f.bool is not None:
        return bool_v(x) if isinstance(x, bool) else _fail(f"field {f.name}: not bool")
    if f.string is not None:
        return str_v(x) if isinstance(x, str) else _fail(f"field {f.name}: not string")
    if f.bytes is not None:
        return bytes_v(x) if isinstance(x, bytes) else _fail(f"field {f.name}: not bytes")
    if f.choice is not None or f.json is not None:
        return from_native(x)
    if f.duration is not None:
        return (
            duration_v(x)
            if isinstance(x, dt.timedelta)
            else _fail(f"field {f.name}: not a duration")
        )
    if f.timestamp is not None:
        return (
            timestamp_v(x)
            if isinstance(x, dt.datetime)
            else _fail(f"field {f.name}: not a timestamp")
        )
    if f.list is not None:
        if not isinstance(x, list):
            return _fail(f"field {f.name}: not a list")
        items = f.list.items
        out: list[Value] = []
        for i, el in enumerate(x):
            item = items[0] if len(items) == 1 else (items[i] if i < len(items) else None)
            out.append(from_native(el) if item is None else canonical_value(item, el))
        return list_v(*out)
    if f.object is not None:
        if not isinstance(x, dict):
            return _fail(f"field {f.name}: not an object")
        sub = f.object.schema
        return from_native(x) if sub is None else canonical_struct(sub, x)
    if f.map is not None:
        if not isinstance(x, dict):
            return _fail(f"field {f.name}: not a map")
        vs = f.map.value_schema
        fields: dict[str, Value] = {}
        for key, el in x.items():
            if vs is not None and isinstance(el, dict):
                fields[key] = canonical_struct(vs, el)
            else:
                fields[key] = from_native(el)
        return struct_v(fields)
    # Computed / OneOf / Ref canonicalize structurally; the engine resolves
    # through their target schemas instead.
    return from_native(x)


def canonical_struct(s: Schema, m: NativeStruct) -> Value:
    fields: dict[str, Value] = {}
    by_name = {f.name: f for f in s.fields}
    for key, el in m.items():
        fld = by_name.get(key)
        fields[key] = from_native(el) if fld is None else canonical_value(fld, el)
    return struct_v(fields)

File: go/schemapb/value.go
package schemapb

import (
	"fmt"
	"math"
	"time"

	"google.golang.org/protobuf/types/known/durationpb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// This file is the single conversion point between the three value worlds:
//
//   wire    — *Value / *StructValue (the typed protobuf contract)
//   native  — plain Go values the engine and CEL operate on
//   schema  — the canonical wire variant dictated by a field's declared kind
//
// Native model (what the engine sees):
//
//	Null      -> nil                 Bool      -> bool
//	Int32/64  -> int64               UInt32/64 -> uint64
//	Float/Double -> float64          String    -> string
//	Bytes     -> []byte              Enum      -> int64
//	Duration  -> time.Duration       Timestamp -> time.Time
//	List      -> []any               Object/Map -> map[string]any
//	Json      -> any of the above
//
// All signed integers widen to int64 and floats to float64 in the native
// model (CEL's int/uint/double are 64-bit); the declared field kind narrows
// them back to the canonical wire variant at the boundary (CanonicalValue).

// =============================================================================
// Constructors (wire values)
// =============================================================================

// NullV returns a wire null.
func NullV() *Value { return &Value{Kind: &Value_NullValue{}} }

// BoolV returns a wire bool.
func BoolV(v bool) *Value { return &Value{Kind: &Value_BoolValue{BoolValue: v}} }

// Int32V returns a wire int32.
func Int32V(v int32) *Value { return &Value{Kind: &Value_Int32Value{Int32Value: v}} }

// Int64V returns a wire int64.
func Int64V(v int64) *Value { return &Value{Kind: &Value_Int64Value{Int64Value: v}} }

// UInt32V returns a wire uint32.
func UInt32V(v uint32) *Value { return &Value{Kind: &Value_Uint32Value{Uint32Value: v}} }

// UInt64V returns a wire uint64.
func UInt64V(v uint64) *Value { return &Value{Kind: &Value_Uint64Value{Uint64Value: v}} }

// FloatV returns a wire float.
func FloatV(v float32) *Value { return &Value{Kind: &Value_FloatValue{FloatValue: v}} }

// DoubleV returns a wire double.
func DoubleV(v float64) *Value { return &Value{Kind: &Value_DoubleValue{DoubleValue: v}} }

// StrV returns a wire string.
func StrV(v string) *Value { return &Value{Kind: &Value_StringValue{StringValue: v}} }

// BytesV returns a wire bytes value.
func BytesV(v []byte) *Value { return &Value{Kind: &Value_BytesValue{BytesValue: v}} }

// DurationV returns a wire duration.
func DurationV(d time.Duration) *Value {
	return &Value{Kind: &Value_DurationValue{DurationValue: durationpb.New(d)}}
}

// TimestampV returns a wire timestamp.
func TimestampV(t time.Time) *Value {
	return &Value{Kind: &Value_TimestampValue{TimestampValue: timestamppb.New(t)}}
}

// ListV returns a wire list of the given items.
func ListV(items ...*Value) *Value {
	return &Value{Kind: &Value_ListValue{ListValue: &ListValue{Items: items}}}
}

// StructV returns a wire struct with the given fields.
func StructV(fields map[string]*Value) *Value {
	return &Value{Kind: &Value_StructValue{StructValue: &StructValue{Fields: fields}}}
}

// =============================================================================
// Wire -> native
// =============================================================================

// ToGo converts a wire value to its native representation.
//
//nolint:cyclop // flat exhaustive wire-variant dispatch
func (v *Value) ToGo() any {
	switch k := v.GetKind().(type) {
	case nil, *Value_NullValue:
		return nil
	case *Value_BoolValue:
		return k.BoolValue
	case *Value_Int32Value:
		return int64(k.Int32Value)
	case *Value_Int64Value:
		return k.Int64Value
	case *Value_Uint32Value:
		return uint64(k.Uint32Value)
	case *Value_Uint64Value:
		return k.Uint64Value
	case *Value_FloatValue:
		return float64(k.FloatValue)
	case *Value_DoubleValue:
		return k.DoubleValue
	case *Value_StringValue:
		return k.StringValue
	case *Value_BytesValue:
		return k.BytesValue
	case *Value_DurationValue:
		return k.DurationValue.AsDuration()
	case *Value_TimestampValue:
		return k.TimestampValue.AsTime()
	case *Value_ListValue:
		items := k.ListValue.GetItems()
		out := make([]any, len(items))

		for i, it := range items {
			out[i] = it.ToGo()
		}

		return out
	case *Value_StructValue:
		return k.StructValue.ToGo()
	default:
		return nil
	}
}

// ToGo converts a wire struct to a native map. A nil struct yields an empty map.
func (s *StructValue) ToGo() map[string]any {
	fields := s.GetFields()
	out := make(map[string]any, len(fields))

	for name, v := range fields {
		out[name] = v.ToGo()
	}

	return out
}

// =============================================================================
// Native -> wire (best fit, no schema)
// =============================================================================

// FromGo converts a native Go value to a wire value using the best-fitting
// variant (int64 for signed integers, uint64 for unsigned, double for floats).
// Use CanonicalValue when the field kind is known — it picks the exact
// contract variant. Unsupported types return an error.
//
//nolint:cyclop,funlen // flat exhaustive native-type dispatch
func FromGo(x any) (*Value, error) {
	switch t := x.(type) {
	case nil:
		return NullV(), nil
	case *Value:
		return t, nil
	case bool:
		return BoolV(t), nil
	case int:
		return Int64V(int64(t)), nil
	case int8:
		return Int64V(int64(t)), nil
	case int16:
		return Int64V(int64(t)), nil
	case int32:
		return Int64V(int64(t)), nil
	case int64:
		return Int64V(t), nil
	case uint:
		return UInt64V(uint64(t)), nil
	case uint8:
		return UInt64V(uint64(t)), nil
	case uint16:
		return UInt64V(uint64(t)), nil
	case uint32:
		return UInt64V(uint64(t)), nil
	case uint64:
		return UInt64V(t), nil
	case float32:
		return DoubleV(float64(t)), nil
	case float64:
		return DoubleV(t), nil
	case string:
		return StrV(t), nil
	case []byte:
		return BytesV(t), nil
	case time.Duration:
		return DurationV(t), nil
	case time.Time:
		return TimestampV(t), nil
	case []any:
		items := make([]*Value, len(t))

		for i, el := range t {
			v, err := FromGo(el)
			if err != nil {
				return nil, fmt.Errorf("list[%d]: %w", i, err)
			}

			items[i] = v
		}

		return ListV(items...), nil
	case map[string]any:
		fields := make(map[string]*Value, len(t))

		for name, el := range t {
			v, err := FromGo(el)
			if err != nil {
				return nil, fmt.Errorf("%s: %w", name, err)
			}

			fields[name] = v
		}

		return StructV(fields), nil
	default:
		return nil, fmt.Errorf("schemapb: cannot convert %T to Value", x)
	}
}

// MustFromGo is FromGo that panics on an unsupported type. For literals and
// tests; prefer FromGo for runtime data.
func MustFromGo(x any) *Value {
	v, err := FromGo(x)
	if err != nil {
		panic(err)
	}

	return v
}

// StructFromGo converts a native map to a wire struct.
func StructFromGo(m map[string]any) (*StructValue, error) {
	fields := make(map[string]*Value, len(m))

	for name, el := range m {
		v, err := FromGo(el)
		if err != nil {
			return nil, fmt.Errorf("%s: %w", name, err)
		}

		fields[name] = v
	}

	return &StructValue{Fields: fields}, nil
}

// MustStructFromGo is StructFromGo that panics on an unsupported type. For
// literals and tests; prefer StructFromGo for runtime data.
func MustStructFromGo(m map[string]any) *StructValue {
	st, err := StructFromGo(m)
	if err != nil {
		panic(err)
	}

	return st
}

// =============================================================================
// Native numeric coercion helpers (shared by validate/compute)
// =============================================================================

// asInt64 extracts a signed integer from any native numeric representation.
// Floats convert only when integral; uints only when in range.
func asInt64(x any) (int64, bool) {
	switch n := x.(type) {
	case int64:
		return n, true
	case int:
		return int64(n), true
	case int32:
		return int64(n), true
	case uint64:
		if n <= math.MaxInt64 {
			return int64(n), true
		}
	case float64:
		if n == math.Trunc(n) && n >= math.MinInt64 && n <= math.MaxInt64 {
			return int64(n), true
		}
	case float32:
		return asInt64(float64(n))
	}

	return 0, false
}

// asUint64 extracts an unsigned integer from any native numeric representation.
func asUint64(x any) (uint64, bool) {
	switch n := x.(type) {
	case uint64:
		return n, true
	case uint32:
		return uint64(n), true
	case int64:
		if n >= 0 {
			return uint64(n), true
		}
	case int:
		if n >= 0 {
			return uint64(n), true
		}
	case float64:
		if n == math.Trunc(n) && n >= 0 && n <= math.MaxUint64 {
			return uint64(n), true
		}
	case float32:
		return asUint64(float64(n))
	}

	return 0, false
}

// asFloat64 extracts a float from any native numeric representation.
func asFloat64(x any) (float64, bool) {
	switch n := x.(type) {
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int64:
		return float64(n), true
	case int:
		return float64(n), true
	case int32:
		return float64(n), true
	case uint64:
		return float64(n), true
	case uint32:
		return float64(n), true
	}

	return 0, false
}

// =============================================================================
// Canonical form (field kind -> exact wire variant)
// =============================================================================

// CanonicalValue converts a native value to the exact wire variant the field's
// declared kind mandates (the contract's canonical form): an Int64 field's
// value is always int64_value, a Float field's always float_value, and so on.
// A value that cannot represent the kind (wrong type, out of range) returns an
// error; the validator reports such values as TYPE_MISMATCH before this point.
//
//nolint:gocognit,cyclop,gocyclo,funlen // flat exhaustive kind dispatch
func CanonicalValue(f *Schema_Field, x any) (*Value, error) {
	if x == nil {
		return NullV(), nil
	}

	switch f.GetKind().(type) {
	case *Schema_Field_Float_:
		n, ok := asFloat64(x)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not numeric", f.GetName(), x)
		}

		return FloatV(float32(n)), nil
	case *Schema_Field_Double_:
		n, ok := asFloat64(x)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not numeric", f.GetName(), x)
		}

		return DoubleV(n), nil
	case *Schema_Field_Int32_:
		n, ok := asInt64(x)
		if !ok || n < math.MinInt32 || n > math.MaxInt32 {
			return nil, fmt.Errorf("field %s: %v does not fit int32", f.GetName(), x)
		}

		return Int32V(int32(n)), nil
	case *Schema_Field_Int64_:
		n, ok := asInt64(x)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not an integer", f.GetName(), x)
		}

		return Int64V(n), nil
	case *Schema_Field_Uint32:
		n, ok := asUint64(x)
		if !ok || n > math.MaxUint32 {
			return nil, fmt.Errorf("field %s: %v does not fit uint32", f.GetName(), x)
		}

		return UInt32V(uint32(n)), nil
	case *Schema_Field_Uint64:
		n, ok := asUint64(x)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not an unsigned integer", f.GetName(), x)
		}

		return UInt64V(n), nil
	case *Schema_Field_Bool_:
		b, ok := x.(bool)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not bool", f.GetName(), x)
		}

		return BoolV(b), nil
	case *Schema_Field_String_:
		s, ok := x.(string)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not string", f.GetName(), x)
		}

		return StrV(s), nil
	case *Schema_Field_Bytes_:
		b, ok := x.([]byte)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not bytes", f.GetName(), x)
		}

		return BytesV(b), nil
	case *Schema_Field_Choice_:
		// Choice values are typed by their options; canonical form is the
		// best-fit variant of the native value.
		return FromGo(x)
	case *Schema_Field_Duration_:
		d, ok := x.(time.Duration)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not a duration", f.GetName(), x)
		}

		return DurationV(d), nil
	case *Schema_Field_Timestamp_:
		t, ok := x.(time.Time)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not a timestamp", f.GetName(), x)
		}

		return TimestampV(t), nil
	case *Schema_Field_List_:
		arr, ok := x.([]any)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not a list", f.GetName(), x)
		}

		items := make([]*Value, len(arr))

		for i, el := range arr {
			var v *Value

			var err error
			if item := listItemDef(f.GetList(), i); item != nil {
				v, err = CanonicalValue(item, el)
			} else {
				v, err = FromGo(el)
			}

			if err != nil {
				return nil, fmt.Errorf("field %s[%d]: %w", f.GetName(), i, err)
			}

			items[i] = v
		}

		return ListV(items...), nil
	case *Schema_Field_Object_:
		m, ok := x.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not an object", f.GetName(), x)
		}

		return canonicalStruct(f.GetObject().GetSchema(), m, f.GetName())
	case *Schema_Field_Map_:
		m, ok := x.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("field %s: %T is not a map", f.GetName(), x)
		}

		vs := f.GetMap().GetValueSchema()
		fields := make(map[string]*Value, len(m))

		for key, el := range m {
			if vs != nil {
				em, isObj := el.(map[string]any)
				if !isObj {
					return nil, fmt.Errorf("field %s.%s: %T is not an object", f.GetName(), key, el)
				}

				v, err := canonicalStruct(vs, em, f.GetName()+"."+key)
				if err != nil {
					return nil, err
				}

				fields[key] = v

				continue
			}

			v, err := FromGo(el)
			if err != nil {
				return nil, fmt.Errorf("field %s.%s: %w", f.GetName(), key, err)
			}

			fields[key] = v
		}

		return StructV(fields), nil
	case *Schema_Field_Json_:
		return FromGo(x)
	case *Schema_Field_Computed_:
		// Computed values canonicalize by the declared result type.
		return canonicalResult(f.GetComputed().GetResult(), x)
	default:
		// OneOf / Ref values canonicalize structurally (variant/def schemas are
		// resolved by the engine, which canonicalizes through them instead).
		return FromGo(x)
	}
}

// canonicalStruct canonicalizes a native map against a schema's declared
// fields; keys without a declared field fall back to best-fit conversion.
func canonicalStruct(s *Schema, m map[string]any, path string) (*Value, error) {
	fields := make(map[string]*Value, len(m))

	for key, el := range m {
		var fld *Schema_Field

		for _, f := range s.GetFields() {
			if f.GetName() == key {
				fld = f

				break
			}
		}

		var v *Value

		var err error
		if fld != nil {
			v, err = CanonicalValue(fld, el)
		} else {
			v, err = FromGo(el)
		}

		if err != nil {
			return nil, fmt.Errorf("%s.%s: %w", path, key, err)
		}

		fields[key] = v
	}

	return StructV(fields), nil
}

// canonicalResult canonicalizes a computed result by its declared ResultType.
//
//nolint:cyclop // flat exhaustive ResultType dispatch
func canonicalResult(rt Schema_Field_ResultType, x any) (*Value, error) {
	switch rt {
	case Schema_Field_RESULT_TYPE_DOUBLE:
		if n, ok := asFloat64(x); ok {
			return DoubleV(n), nil
		}
	case Schema_Field_RESULT_TYPE_INT64:
		if n, ok := asInt64(x); ok {
			return Int64V(n), nil
		}
	case Schema_Field_RESULT_TYPE_UINT64:
		if n, ok := asUint64(x); ok {
			return UInt64V(n), nil
		}
	case Schema_Field_RESULT_TYPE_BOOL:
		if b, ok := x.(bool); ok {
			return BoolV(b), nil
		}
	case Schema_Field_RESULT_TYPE_STRING:
		if s, ok := x.(string); ok {
			return StrV(s), nil
		}
	case Schema_Field_RESULT_TYPE_DURATION:
		if d, ok := x.(time.Duration); ok {
			return DurationV(d), nil
		}
	case Schema_Field_RESULT_TYPE_TIMESTAMP:
		if t, ok := x.(time.Time); ok {
			return TimestampV(t), nil
		}
	case Schema_Field_RESULT_TYPE_BYTES:
		if b, ok := x.([]byte); ok {
			return BytesV(b), nil
		}
	default:
		return FromGo(x)
	}

	return nil, fmt.Errorf("computed result %T does not match declared type %v", x, rt)
}
